src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -17,123 +21,170 @@ class MSMonitor(MSCore):
                 # move the request status staging -> staged
-                elif self.pileupTransfersCompleted(transferRecords):
+                elif self.completion(transfers, pileup=True):
                     self.logger.debug(
-                        "+++ request %s pileup transfers are completed", req)
-                    self.change(req, 'staged', '+++ monit')
+                        "+++ request %s pileup transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
                 # transfers not completed
                 # just update the database with their completion
                 else:
                     self.logger.debug(
-                        "+++ request %s transfers are %s completed",
-                        req, completion)
-                    # for us workflow name is the same as request name
-                    wname = reqName
-                    requestStatuses[wname] = transferRecords
-            self.updateTransferInfo(requestStatuses)
+                        "+++ request %s transfers are completed", reqName)
+            self.updateTransferInfo(transferRecords, completedWorkflows)
+            # finally, update statuses for requests
+            for reqName, status in reqStatus.items():
+                self.change(reqName, status, '+++ monit')
         except Exception as err:  # general error
             self.logger.exception('+++ monit error: %s', str(err))
 
-    def getTransferRecords(self, reqName):
+    def completion(self, transfers, pileup=None):
+        """
+        Helper function to calculate completion of given records
+        :param transfers: list of transfers records
+        :param pileup: check pileup completion (optional)
+        :return: completion status
+        """
+        # check completion of all transfers
+        statuses = []
+        transferTypes = ['primary', 'secondary']
+        for rec in transfers:
+            campaign = rec['campaignName']
+            cdict = self.campaigns[campaign]
+            thr = cdict.get('PartialCopy', 0)
+            if pileup and rec['dataType'] not in transferTypes:
+                continue
+            if rec['completion'] >= thr:
+                status = 1
+            else:
+                status = 0
+            statuses.append(status)
+        return True if sum(statuses) == len(transfers) else False
+
+    def updateTransferInfo(self, transferRecords, completedWorkflows):
+        """
+        Update transfer records in backend
+        :param transferRecords: list of transfer records
+        :param completedWorkflows: list of completed workflow names to be removed from backend
+        """
+        tstamp = time.time()
+        for doc in transferRecords:
+            if doc['workflowName'] in completedWorkflows:
+                continue  # we skip completed workflows
+            transfers = []
+            for rec in doc.get('transfers', []):
```

Comment:
The transfer ID is set in stone by now and it's not responsibility of the MSMonitor to change/update it (well, not at this stage where we don't have transfers troubleshooting). It will only update the transfer completion and `lastUpdate` timestamp.

Reply:
I'm not sure that it is fixed? It is up to upstream system (PhEDEx or Rucio) to decide when/how to transfer blocks. Therefore I would expect that transfer ids set will/may change over time. There is no hard to update them either if their are fixed since they will never change. Please clarify your thoughts on this before requesting the change.

Reply:
PhEDEx/Rucio are simply making the data movement that someone/something ordered them to do. So once a request ID is created, PhEDEx/Rucio will consider that and start the data movement. Of course a given data is not associated to only 1 request, but what matters to us is what MicroService is requesting, we can't control what others request.